torch/jit/_check.py

```diff
@@ -0,0 +1,214 @@
+    setting an instance-level attribute to ``[]`` (for ``List``),
+    ``{}`` for ``Dict``), or ``None`` (for ``Optional``) isn't enough
+    information for us to properly initialize that attribute.
+
+    An object of this class can walk a given ``nn.Module``'s AST and
+    determine if it meets our requirements or not.
+
+    Known limitation: We can only check the AST nodes for certain
+    constructs; we can't ``eval`` arbitrary expressions. This means
+    that function calls, class instantiations, and complex expressions
+    that resolve to one of the "empty" values specified above will NOT
+    be flagged as problematic.
+
+    Example:
+
+        .. code-block:: python
+
+            class M(torch.nn.Module):
+                def fn(self):
+                    return []
+
+                def __init__(self):
+                    super().__init__()
+                    self.x: List[int] = []
+
+                def forward(self, x: List[int]):
+                    self.x = x
+                    return 1
+
+        The above code will pass the ``AttributeTypeIsSupportedChecker``
+        check since we have a function call in ``__init__``. However,
+        it will still fail later with the ``RuntimeError`` "Tried to set
+        nonexistent attribute: x. Did you forget to initialize it in
+        __init__()?".
+
+    Args:
+        nn_module - The instance of ``torch.nn.Module`` whose
+            ``__init__`` method we wish to check
+    """
+
+    def check(self, nn_module: torch.nn.Module) -> None:
+
+        # Check if we have a Python version <3.8
+        self.using_deprecated_ast: bool = sys.version_info < (3, 8)
+
+        source_lines = textwrap.dedent(inspect.getsource(nn_module.__class__.__init__))
+
+        # This AST only contains the `__init__` method of the nn.Module
+        init_ast = ast.parse(source_lines)
+
+        self.visit(init_ast)
+
+    def visit_AnnAssign(self, node):
+        """
+        Visit an AnnAssign node in an ``nn.Module``'s ``__init__``
+        method and see if it conforms to our attribute annotation rules.
+        """
+        # If we have a class-level attribute (instance-level attributes
+        # are instances of ast.Attribute)
+        if isinstance(node.target, ast.Name):
+            return
+
+        # TODO @ansley: add `Union` once landed
+
+        # NB: Even though `Tuple` is a "container", we don't want to
+        # check for it here. `Tuple` functions as an type with an
+        # "infinite" number of subtypes, in the sense that you can have
+        # `Tuple[())]`, `Tuple[T1]`, `Tuple[T2]`, `Tuple[T1, T2]`,
+        # `Tuple[T2, T1]` and so on, and none of these subtypes can be
+        # used in place of the other. Therefore, assigning an empty
+        # tuple in `__init__` CORRECTLY means that that variable
+        # cannot be reassigned later to a non-empty tuple. Same
+        # deal with `NamedTuple`
+
+        containers = {"List", "Dict", "Optional"}
```

Comment:
Note that matching on names is not necessarily reliable, since `from typing import List as foo`  is possible. But this seems fine, probably no one does that.